### DISCUSSION TITLE
Read String but just showing the first character

### DISCUSSION THREAD

**Discussion by maristyar** [aldas/modbus-tcp-client#162] [Q&A]
Have followed the instruction to read the string using this code, but just showing the first char. Anyone can help me?
![Screenshot (50)](https://github.com/aldas/modbus-tcp-client/assets/60589302/80d7fa69-70ec-43e1-9ae6-19a7f77bb599)

```php
    <?php
    session_start();
    require __DIR__ . '/vendor/autoload.php';

    use ModbusTcpClient\Composer\Address;
    use ModbusTcpClient\Composer\Read\ReadCoilsBuilder;
    use ModbusTcpClient\Composer\Read\ReadRegistersBuilder;
    use ModbusTcpClient\Composer\Write\WriteRegistersBuilder;
    use ModbusTcpClient\Network\NonBlockingClient;
    use ModbusTcpClient\Utils\Endian;
    
    Endian::$defaultEndian = Endian::BIG_ENDIAN_LOW_WORD_FIRST; 

    $addressPLC = 'tcp://127.0.0.1:502';

    // READ COIL DATA
    $fc1 = ReadCoilsBuilder::newReadCoils($addressPLC)
        ->coil(0, 'TRUCK_01_DI_PB_START')
        ->coil(1, 'TRUCK_01_DI_PB_STOP')
        ->coil(2, 'TRUCK_01_DI_ESD')
        ->coil(10, 'TRUCK_01_DO_START_STATUS')
        ->coil(11, 'TRUCK_01_DO_STOP_STATUS')
        ->build();

        $responseFC1 = (new NonBlockingClient(['readTimeoutSec' => 0.2]))->sendRequests($fc1);
        $dataModbusFC1 = $responseFC1->getData();
        $errorModbusFC1 = $responseFC1->getErrors();

    // READ REGISTER DATA
    $fc3 = ReadRegistersBuilder::newReadHoldingRegisters($addressPLC)
        ->float(0, 'dataFloat')
        ->int16(2, 'dataInteger')
        ->string(
            4, 
            6, 
            'dataString',
            function ($value, $address, $response) {
                return $value;
            },
            function (\Exception $exception, Address $address, $response) {
                return $address->getType() === Address::TYPE_STRING ? '' : null;
            }
        )
        ->build();

        $responseFC3 = (new NonBlockingClient(['readTimeoutSec' => 0.2]))->sendRequests($fc3);
        $dataModbusFC3 = $responseFC3->getData();
        $errorModbusFC3 = $responseFC3->getErrors();


    $dataModbus = array_merge($dataModbusFC1, $dataModbusFC3);
    $errorModbus = array_merge($errorModbusFC1, $errorModbusFC3);
```

**Comment by aldas** [comment]
You could try to debug log what is send and what is received. Maybe it is endianess problem.  Maybe charset

```php
<?php
session_start();
require __DIR__ . '/vendor/autoload.php';
require __DIR__ . '/logger.php';

use ModbusTcpClient\Composer\Read\ReadRegistersBuilder;
use ModbusTcpClient\Network\NonBlockingClient;
use ModbusTcpClient\Utils\Endian;

Endian::$defaultEndian = Endian::BIG_ENDIAN_LOW_WORD_FIRST;

$addressPLC = 'tcp://127.0.0.1:502';

// READ REGISTER DATA
$fc3 = ReadRegistersBuilder::newReadHoldingRegisters($addressPLC)
    ->string(4, 6, 'dataString')
    ->build();

try {
    $responseFC3 = (new NonBlockingClient([
        'readTimeoutSec' => 0.2,
        'logger' => new EchoLogger()
    ]))->sendRequests($fc3);
    
    $dataModbusFC3 = $responseFC3->getData();
    $errorModbusFC3 = $responseFC3->getErrors();

    print_r($dataModbusFC3);
} catch (Exception $exception) {
    echo 'An exception occurred' . PHP_EOL;
    echo $exception->getMessage() . PHP_EOL;
    echo $exception->getTraceAsString() . PHP_EOL;
}
```

That logger source is here https://github.com/aldas/modbus-tcp-client/blob/master/examples/logger.php

**Reply by maristyar** [reply]
aldas, there is a problem like this
"Fatal error: Uncaught Error: Interface "Psr\Log\LoggerInterface" not found in C:\xampp\htdocs\modbus\real\vendor\aldas\modbus-tcp-client\examples\logger.php:8 Stack trace: #0 C:\xampp\htdocs\modbus\real\data.php(4): require() #1 {main} thrown in C:\xampp\htdocs\modbus\real\vendor\aldas\modbus-tcp-client\examples\logger.php on line 8"

What do I have to do?

**Reply by aldas** [reply]
You can get it with composer https://packagist.org/packages/psr/log

**Reply by maristyar** [reply]
Hi aldas, I found the responds like this

2024-05-09 05:29:28.806266: [debug] msg: "Connected" 2024-05-09 05:29:28.806420: [debug] msg: "Data sent", data: array ( 1 => 'c81a00000006000300040003', ) 2024-05-09 05:29:28.806474: [debug] msg: "Polling data" 2024-05-09 05:29:28.806488: [debug] msg: "Stream 34 @ index: 0 received data: ", data: array ( 1 => 'c81a0000000900030600610064006d', ) Array ( [dataString] => a )

I think the responds data was correct. But how to display it?

**Reply by aldas** [reply]
So the response is `c81a0000000900030600610064006d` 
Response FC3 structure can be seen here
https://github.com/aldas/modbus-tcp-client/blob/b18323bf5bb2abb531d97d6f385bafb3a6c0c658/src/Packet/ModbusFunction/ReadHoldingRegistersResponse.php#L18-L27

so the raw string is `00610064006d` in hex and in binary `�a�d�m`. It seems that every character is 2 bytes (1 register).  This does not seem to be ASCII.  Are you setting each character in individual register - and that is why it is 2 bytes per character?

```php
    public function testString()
    {
        $result = mb_convert_encoding("\x00\x61\x00\x64\x00\x6d", 'UTF-8', 'UTF-16');
        $this->assertEquals('adm', $result);
    }
```

seems to give satisfactory result but it could be just an coincidentally. 

-----

How many characters you string should have? is it 6? I am asking this because `->String` works with ASCII strings that are 1 character per byte so in modbus sense 1 register/word can contain 2 characters. 

------

I think I should add some kind of raw bytes method to `ReadRegistersBuilder` so this kind of byte manipulation could be done with callback.

**Reply by aldas** [reply]
note: for hacking 1-4 character lengths we could use int64, but this is a hack more than a solution
```php
                ->int64( // int64 is 8 bytes, but we want 6 (3*2byte characters)
                    4,
                    'dataString',
                    function ($value, $address, $response) {
                        // https://www.php.net/manual/en/function.pack.php
                        // P	unsigned long long (always 64 bit, little endian byte order)
                        $binary = pack('P', $value);

                        // cut of last 2 bytes from int64
                        $rawText = substr($binary, 0, 6);
                        // convert it into proper utf-8 string
                        return mb_convert_encoding($rawText, 'UTF-8', 'UTF-16');
                    },
                    null,
                    Endian::LITTLE_ENDIAN,
                    )
```

**Reply by maristyar** [reply]
That's great, aldas. By using your last code it is showing the correct string. 

 `2024-05-09 07:06:49.510344: [debug] msg: "Connected" 2024-05-09 07:06:49.510429: [debug] msg: "Data sent", data: array ( 1 => 'bf5100000006000300040004', ) 2024-05-09 07:06:49.510519: [debug] msg: "Polling data" 2024-05-09 07:06:49.510537: [debug] msg: "Stream 33 @ index: 0 received data: ", data: array ( 1 => 'bf510000000b00030800610064006d0000', ) Array ( [dataString] => adm )'
 
But, it is just my trial. The actual length of string may be 20 characters or more. And, the PLC M221 only can send string char-per-char OR per-individual-register. Any suggestion from you how to read complete string?

**Reply by maristyar** [reply]
Hi, aldas. I have same issue to write the string like this. From the register %MW4 until %MW7, I send "admi" but the my plc receive the foreign char like this. Hope you can help me. Thanks.

![Screenshot (52)](https://github.com/aldas/modbus-tcp-client/assets/60589302/033e15d4-ed0b-4a0b-86d7-178bb7a24d30)

**Reply by aldas** [reply]
This is because this library handles (reading/writing) strings as 1character=1byte and fits 2 characters (following each other) into 1register/word. 

What you could do for writing is to break your  string into characters and use PHP ord (https://www.php.net/manual/en/function.ord.php) function to get (unsigned) integer value for that character and send them as uint16 value - 1 uint16 per register.

**Reply by maristyar** [reply]
Aldas, one more question. The default Endian is BIG_ENDIAN_LOW_WORD_FIRST. But, how to change the Endian to LITTLE_ENDIAN just for write string start from register 4 like this code:

```php
$fc4 = WriteRegistersBuilder::newWriteMultipleRegisters($addressPLC)
    ->float(0, 2.2)
    ->int16(2, 8)
    ->string(4, "fajkafksafkaj118391jkfjkaj9%^^*", 30)
    ->build();

    $responseFC4 = (new NonBlockingClient(['readTimeoutSec' => 3]))->sendRequests($fc4);
```
